### PR TITLE
Convert get_for_path to get_translatable

### PR DESCRIPTION
**Pull request by phlax**
- replaces Unit.get_for_path with Unit.get_translatable
- avoids use of path or regex wherever possible

### CODE REVIEW

**Comment by unho** [comment]
lgtm

**Comment by unho** [comment]
Shouldn't both commits be squashed? Is it really necessary to alter the order of the parameters?

**Comment by phlax** [comment]
> Is it really necessary to alter the order of the parameters?

this method is useful for filtering out the template dirs for a user without any other filter - so it seems to make sense to switch the args and make the path optional.

**Comment by phlax** [comment]
> Shouldn't both commits be squashed?

maybe - but i expected some conversation about the second commit - so kept separate

File: pootle/apps/pootle_store/models.py [hunk 1]
```diff
@@ -235,48 +235,27 @@ def get_for_user(self, user):
...
-                units_path = ''.join(['/%/', proj, '/', filename])
-            else:
-                units_path = ''.join(['/%/', proj, '/%'])
-        # /projects/translate/*
-        elif lang is None and proj is None:
-            units_path = '/%'
-        # /<lang_code>/<project_code>/translate/*
-        # /<lang_code>/translate/*
-        else:
-            units_path = ''.join([pootle_path, '%'])
```
Comment (unho) [inline pootle/apps/pootle_store/models.py]: How do we get these by using the regex?

Comment (phlax) [inline pootle/apps/pootle_store/models.py]: atm we dont - ill add a WIP header

File: pootle/core/views.py [hunk 2]
```diff
@@ -49,6 +49,102 @@
...
+    @property
+    def filename(self):
+        return self.kwargs.get("filename", "") or ""
+
+    @property
+    def language_code(self):
+        return self.kwargs.get("language_code", None) or None
+
+    @cached_property
+    def path(self):
```
Comment (dwaynebailey) [inline pootle/core/views.py]: Codacy Issue found: [An attribute defined in pootle.core.views line 108 hides this method](https://www.codacy.com/app/dwayne/pootle/file/1742333681/issues/source?bid=1990595&fileBranchId=3039950#l134)

Comment (phlax) [inline pootle/core/views.py]: at last! some useful output from codacy 8/

File: pootle/apps/pootle_store/models.py (hunk 1 above)
Comment (unho) [inline pootle/apps/pootle_store/models.py]: Is this sorted out now?

Comment (phlax) [inline pootle/apps/pootle_store/models.py]: yep - i believe so

File: pootle/core/views.py [hunk 3]
```diff
@@ -49,6 +50,115 @@
...
+    project_regex = r"[a-zA-Z0-9\-\_]*"
+
+    @property
+    def regex(self):
+        if not (self.project_code or self.language_code):
+            return None
+        regex = r"^/"
+
+        if self.language_code:
+            regex = r"%s%s/" % (regex, self.language_code)
```
Comment (unho) [inline pootle/core/views.py]: I find it simpler to understand like this:

```python
            regex += r"%s/" % self.language_code
```

Comment (phlax) [inline pootle/core/views.py]: not sure  i agree

File: pootle/core/views.py [hunk 4]
```diff
@@ -49,6 +50,115 @@
...
+        return self.kwargs.get("language_code", None) or None
+
+    @cached_property
+    def path(self):
+        if not self.request_path:
+            return self.request_path
+        path = self._extracted_vfolder_and_path[1]
+        parts = path.split("/")
+        if "." in parts[-1]:
+            return "/%s" % path.strip("/")
```
Comment (unho) [inline pootle/core/views.py]: Simpler:

```python
            return path.rstrip("/")
```

File: pootle/core/views.py (hunk 3 above)
Comment (unho) [inline pootle/core/views.py]: Perhaps it is less explicit.

File: pootle/core/views.py (hunk 4 above)
Comment (phlax) [inline pootle/core/views.py]: that doesnt guarantee an initial slash - im not sure if it matters tho

Comment (unho) [inline pootle/core/views.py]: All paths have an initial slash. But yes, it is safer not to assume that.

File: pootle/apps/pootle_store/models.py (hunk 1 above)
Comment (unho) [inline pootle/apps/pootle_store/models.py]: I don't see any test for translation paths (paths with `/translate/` on them).

**Comment by julen** [comment]
Unless I'm doing something wrong, testing this on InnoDB shows a massive performance degradation.

  | *master* | *pr/4317*
---- | ----- | -----
/ab/cd/export-view/?search=notebook&sfields=notes,locations |  12 queries run, total 0.125 seconds | 11 queries run, total 13.01 seconds
/ab/cd/export-view/?filter=incomplete | 16 queries run, total 0.191 seconds | 11 queries run, total 6.777 seconds
/ab/export-view/?filter=incomplete | 13 queries run, total 0.262 seconds | 11 queries run, total 6.861 seconds

Here */ab/cd/* is a TP with ~13000 words, and */ab/* contains ~16000 words. The degradation is noticeable in other views too, so this is not limited to the export view.

It might be the case that queries involving regexps do not use an index at all and they have to scan the entire table.

**Comment by phlax** [comment]
@julen i would expect a signifcant performance hit from moving to regexes.

I am not however seeing anything in the order that you are seeing above.

After repeated testing on our staging server against a lang with ~150000 live units the speed goes from ~.6 second to ~.8 second for the relevant regex query

**Comment by phlax** [comment]
...it could be that you are hitting resource limits in your env and this is forcing sorting etc on the disk, otherwise im not really understanding why there would be such a performance hit.

**Comment by phlax** [comment]
@julen would you be able to slowlog your mysql server to find out which queries are taking a long time

**Comment by julen** [comment]
This is in my local machine with far enough resources to handle this operation.

The slow log follows for accessing */ab/cd/export-view/?filter=incomplete*:
```sql
# Query_time: 2.678598  Lock_time: 0.003926 Rows_sent: 1  Rows_examined: 3200834
SELECT COUNT(*) AS `__count`
FROM `pootle_store_unit`
INNER JOIN `pootle_store_store` ON (`pootle_store_unit`.`store_id` = `pootle_store_store`.`id`)
WHERE (`pootle_store_unit`.`state` > -100
       AND NOT (`pootle_store_store`.`pootle_path` LIKE BINARY '/templates/%')
       AND `pootle_store_store`.`pootle_path` REGEXP BINARY '^/ab/cd/'
       AND (`pootle_store_unit`.`state` = 0
            OR `pootle_store_unit`.`state` = 50));

# Query_time: 4.785047  Lock_time: 0.000177 Rows_sent: 1732  Rows_examined: 3200834
SELECT `pootle_store_unit`.`id`,
       `pootle_store_unit`.`store_id`,
       `pootle_store_unit`.`index`,
       `pootle_store_unit`.`unitid`,
       `pootle_store_unit`.`unitid_hash`,
       `pootle_store_unit`.`source_f`,
       `pootle_store_unit`.`source_hash`,
       `pootle_store_unit`.`source_wordcount`,
       `pootle_store_unit`.`source_length`,
       `pootle_store_unit`.`target_f`,
       `pootle_store_unit`.`target_wordcount`,
       `pootle_store_unit`.`target_length`,
       `pootle_store_unit`.`developer_comment`,
       `pootle_store_unit`.`translator_comment`,
       `pootle_store_unit`.`locations`,
       `pootle_store_unit`.`context`,
       `pootle_store_unit`.`state`,
       `pootle_store_unit`.`revision`,
       `pootle_store_unit`.`creation_time`,
       `pootle_store_unit`.`mtime`,
       `pootle_store_unit`.`submitted_by_id`,
       `pootle_store_unit`.`submitted_on`,
       `pootle_store_unit`.`commented_by_id`,
       `pootle_store_unit`.`commented_on`,
       `pootle_store_unit`.`reviewed_by_id`,
       `pootle_store_unit`.`reviewed_on`,
       `pootle_store_unit`.`priority`,
       `pootle_store_store`.`id`,
       `pootle_store_store`.`file`,
       `pootle_store_store`.`parent_id`,
       `pootle_store_store`.`translation_project_id`,
       `pootle_store_store`.`pootle_path`,
       `pootle_store_store`.`name`,
       `pootle_store_store`.`file_mtime`,
       `pootle_store_store`.`state`,
       `pootle_store_store`.`creation_time`,
       `pootle_store_store`.`last_sync_revision`,
       `pootle_store_store`.`obsolete`
FROM `pootle_store_unit`
INNER JOIN `pootle_store_store` ON (`pootle_store_unit`.`store_id` = `pootle_store_store`.`id`)
WHERE (`pootle_store_unit`.`state` > -100
       AND NOT (`pootle_store_store`.`pootle_path` LIKE BINARY '/templates/%')
       AND `pootle_store_store`.`pootle_path` REGEXP BINARY '^/ab/cd/'
       AND (`pootle_store_unit`.`state` = 0
            OR `pootle_store_unit`.`state` = 50));
```

Note the `BINARY` keyword is being used because we have `utf8_bin` collation (cf. #3659). That doesn't affect performance though — and yeah, I checked temporarily switching to `utf8_general_ci` just in case.

**Comment by phlax** [comment]
@julen so im minded that we can probably get rid of the regexes altogether by not searching by path, we already have the parsed lang/proj/dir_path/filename - so there isnt really that much point to reassembling them in a regex/path.

however, im really keen to find out whats causing such significant differences in our 2 test environments.

**Comment by unho** [comment]
lgtm

**Comment by ta2-1** [comment]
@phlax, tested on translate and export views, it's working not worse. lgtm.

**Comment by julen** [comment]
Wonderful, this now works great. Well done @phlax! :+1:

**Pull request merged**